codex/stores/maintenance.nim

```diff
@@ -10,92 +10,323 @@
-    offset = self.offset
+  await self.metaDs.modify(key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        if currDatasetMd.expiry != datasetMd.expiry or currDatasetMd.manifestsCids != datasetMd.manifestsCids:
+          raise newException(CatchableError, "Change in expiry detected, interrupting maintenance for dataset with treeCid " & $treeCid)
+
+        if currDatasetMd.checkpoint.progress > datasetMd.checkpoint.progress:
+          raise newException(CatchableError, "Progress should be increasing only, treeCid " & $treeCid)
+
+        if currDatasetMd.leavesCount <= datasetMd.checkpoint.progress:
+          DatasetMetadata.none
+        else:
+          datasetMd.some
+      else:
+        raise newException(CatchableError, "Metadata for dataset with treeCid " & $treeCid & " not found")
   )
 
-  without iter =? expirations, err:
-    trace "Unable to obtain blockExpirations iterator from repoStore"
-    return
+proc deleteBatch(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  var datasetMd = datasetMd
+
+  datasetMd.checkpoint.timestamp = self.clock.now
+
+  if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+    return failure(err)
+
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    startingIndex = datasetMd.checkpoint.progress
+
+  trace "Deleting a batch of blocks", size = self.batchSize
+
+  var index = datasetMd.checkpoint.progress
+  while (index < datasetMd.checkpoint.progress + self.batchSize) and
+          (index < datasetMd.leavesCount):
+    if err =? (await self.blockStore.delBlock(treeCid, index)).errorOption:
+      error "Error deleting a block", msg = err.msg, index
+
+    index.inc
+
+    await sleepAsync(50.millis) # cooperative scheduling
+
+  if index >= datasetMd.leavesCount:
+    trace "All blocks deleted from a dataset", leavesCount = datasetMd.leavesCount
+
+    for manifestCid in datasetMd.manifestsCids:
+      if err =? (await self.blockStore.delBlock(manifestCid)).errorOption:
+        error "Error deleting manifest", cid = manifestCid
+
+    if err =? (await self.blockStore.delBlock(treeCid)).errorOption:
+      error "Error deleting block", cid = treeCid
+
+    if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+      return failure(err)
+
+    return success()
+  else:
+    datasetMd.checkpoint.progress = index
+    return await self.deleteBatch(treeCid, datasetMd)
+
+proc superviseDatasetDeletion(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[void] {.async.} =
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    expiry = datasetMd.expiry
+    leavesCount = datasetMd.leavesCount
+
+  try:
+    if datasetMd.checkpoint.progress == 0 and datasetMd.checkpoint.timestamp == 0:
+      info "Initiating deletion of a dataset"
```

Comment:
Info is probably way too verbose for this, I would put it behind `trace`.

Reply:
Ditto for the other `info` in the following lines.

Reply:
It's just 2 info logs for the entire dataset. One at the beginning and one at the end of the process. All other logs in the process are on `trace`. I wouldn't call it too verbose, but if you want I can put it.

Reply:
Moving it to `trace`.